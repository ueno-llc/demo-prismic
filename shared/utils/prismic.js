import { RichText } from 'prismic-reactjs';
import _get from 'lodash/get';
import _isPlainObject from 'lodash/isPlainObject';

function linkResolver(doc) {
  if (!doc) {
    return '/';
  }

  switch (doc.type) {
    case 'homepage':
      return '/';
    case 'about':
      return '/about';
    case 'articles':
      return '/articles';
    case 'article':
      return `/articles/${doc.uid}`;
    case 'custom_page':
      return `/${doc.uid}`;
    default:
      return '/';
  }
}

const SHOULD_WARN = process.env.NODE_ENV === 'development';

function warn(...m) {
  if (SHOULD_WARN) {
    console.warn(...m);
  }
}

/**
 * Get string value from object, optionally by path. If `obj` is a string
 * it is returned. If no path is given `obj` is used as the value.
 *
 * @param {Object} The object to query
 * @param {string} [path=undefined] The path of the property to get
 * @param {string} [defaultValue=''] Default value returned if no value is found
 * @returns {*} Resolved value or the empty array if not possible
 */
function getString(obj, path = undefined, defaultValue = '') {
  const value = path ? _get(obj, path) : obj;

  if (!value) {
    return defaultValue;
  }

  if (typeof value === 'string') {
    return value;
  }

  // if we run `asText` on an array that does not have a text key we'll get
  // undefined for that value
  if (Array.isArray(value) && !value.every(i => _isPlainObject(i) && 'text' in i)) {
    return defaultValue;
  }

  try {
    return RichText.asText(value).trim();
  } catch (e) {

    warn(`unable to render field "${path}" as text`, e);
  }

  return defaultValue;
}

/**
 * Get array value from object by path or empty array. If no path is given
 * `obj` is used as the value. If `obj` is an array it is returned.
 *
 * @param {Object} The object to query
 * @param {string} [path=undefined] The path of the property to get
 * @returns {*} Resolved value or the empty array if not possible
 */
function getArray(obj, path = undefined) {
  if (Array.isArray(obj)) {
    return obj;
  }

  const value = path ? _get(obj, path) : obj;

  return Array.isArray(value) ? value : [];
}

/**
 * Get object value from object by path or empty object. If no path is given
 * `obj` is used as the value.
 *
 * @param {Object} The object to query
 * @param {string} [path=undefined] The path of the property to get
 * @returns {*} Resolved value or the empty object if not possible
 */
function getObject(obj, path = undefined) {
  const value = path ? _get(obj, path) : obj;

  return _isPlainObject(value) ? value : {};
}

/**
 * Get richtext value from object by path or defaultValue. If no path is given
 * `obj` is used as the value.
 *
 * @param {Object} The object to query
 * @param {string} [path=undefined] The path of the property to get
 * @param {string} [defaultValue=''] Default value returned if no value is found
 * @returns {*} Resolved value or `defaultValue` if not possible
 */
function getRichtext(obj, path = undefined, defaultValue = '') {
  const value = path ? _get(obj, path) : obj;

  if (!Array.isArray(value)) {
    return defaultValue;
  }

  try {
    return RichText.render(value, linkResolver);
  } catch (e) {
    warn(`unable to render field "${path}" as richtext`, e);
  }

  return defaultValue;
}

export {
  linkResolver,
  getString,
  getObject,
  getArray,
  getRichtext,
};
