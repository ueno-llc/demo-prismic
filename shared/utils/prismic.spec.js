import React from 'react';
import { getString, getArray, getObject, getRichtext } from './prismic';

/* getString */

test('getString for empty object', () => {
  expect(getString({})).toEqual('');
});

test('getString for object with string key', () => {
  expect(getString({ title: 'Test' }, 'title')).toEqual('Test');
});

test('getString for object with string key nested', () => {
  expect(getString({ item: { title: 'Test' } }, 'item.title')).toEqual('Test');
});

test('getString for object with non-existent path', () => {
  expect(getString({ }, 'item.title')).toEqual('');
});

test('getString default value', () => {
  expect(getString({ }, 'item.title', 'default')).toEqual('default');
});

test('getString with string value', () => {
  expect(getString('Title')).toEqual('Title');
});

test('getString with null value', () => {
  expect(getString(null)).toEqual('');
});

test('getString for object with null value', () => {
  expect(getString({ title: null })).toEqual('');
});

test('getString for object with empty array value', () => {
  expect(getString({ title: [] }, 'title')).toEqual('');
});

test('getString for object with array value that does not fit prismic datatype', () => {
  expect(getString({ title: ['foo', 'bar'] }, 'title')).toEqual('');
});

test('getString for object with array value that fits prismic datatype', () => {
  expect(getString({ title: [{ text: 'foo' }, { text: 'bar' }] }, 'title')).toEqual('foo bar');
});

/* getArray */

test('getArray for null', () => {
  expect(getArray(null)).toEqual([]);
});

test('getArray for path of null', () => {
  expect(getArray(null, 'item')).toEqual([]);
});

test('getArray for array', () => {
  expect(getArray([1, 2])).toEqual([1, 2]);
});

test('getArray for array but with path', () => {
  expect(getArray([1, 2], 'item')).toEqual([1, 2]);
});

test('getArray for object with path', () => {
  expect(getArray({ items: [1, 2] }, 'items')).toEqual([1, 2]);
});

test('getArray for object and no path', () => {
  expect(getArray({ items: [1, 2] })).toEqual([]);
});

/* getObject */

test('getObject for null', () => {
  expect(getObject(null)).toEqual({});
});

test('getObject for path of null', () => {
  expect(getObject(null, 'item')).toEqual({});
});

test('getObject for object', () => {
  expect(getObject({ foo: 1 })).toEqual({ foo: 1 });
});

test('getObject for object and path', () => {
  expect(getObject({ item: { title: 'test' } }, 'item')).toEqual({ title: 'test' });
});

test('getObject for object and path with null', () => {
  expect(getObject({ item: null }, 'item')).toEqual({});
});

/* getRichtext */

test('getRichtext for null', () => {
  expect(getRichtext(null)).toEqual('');
});

test('getRichtext for path of null', () => {
  expect(getRichtext(null, 'item')).toEqual('');
});

test('getRichtext for property that is not array', () => {
  expect(getRichtext({ item: 1 }, 'item')).toEqual('');
});

test('getRichtext for property that is an array', () => {
  expect(getRichtext({ items: [1, 2, 3] }, 'item')).toEqual('');
});

test('getRichtext for property that is a richtext array', () => {
  const rich = {
    type: 'paragraph',
    text: 'foo bar',
    spans: [
      { start: 0, end: 3, type: 'strong' },
      { start: 4, end: 7, type: 'em' },
    ],
  };

  expect(getRichtext({ items: [rich] }, 'items')).toMatchSnapshot();
});

