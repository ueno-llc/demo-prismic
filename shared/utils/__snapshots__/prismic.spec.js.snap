// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRichtext for property that is a richtext array 1`] = `
<div>
  <p>
    <strong>
      foo
    </strong>
     
    <em>
      bar
    </em>
  </p>
</div>
`;
